armory/scenarios/poisoning_witches_brew.py

```diff
@@ -0,0 +1,260 @@
+import os
+import copy
+
+import numpy as np
+
+from armory.scenarios.poison import Poison
+from armory.utils.poisoning import FairnessMetrics
+from armory.logs import log
+from armory.utils import config_loading, metrics
+from armory import paths
+
+
+class DatasetPoisonerWitchesBrew:
+    def __init__(
+        self,
+        attack,
+        x_test,
+        y_test,
+        source_class,
+        target_class,
+        trigger_index,
+        data_filepath,
+    ):
+        """
+        Individual source-class triggers are chosen from x_test.  At poison time, the
+        train set is modified to induce misclassification of the triggers as target_class.
+
+        """
+        self.attack = attack
+        self.x_test = x_test
+        self.y_test = y_test
+        self.source_class = source_class
+        self.target_class = target_class
+        self.trigger_index = trigger_index
+        self.data_filepath = data_filepath
+
+    def poison_dataset(self, x_train, y_train, return_index=True):
+        """
+        Return a poisoned version of dataset x, y
+            if return_index, return x, y, index
+        """
+        if len(x_train) != len(y_train):
+            raise ValueError("Sizes of x and y do not match")
+
+        x_trigger = self.x_test[self.trigger_index]
+        if len(x_trigger.shape) == 3:
+            x_trigger = np.expand_dims(x_trigger, axis=0)
+
+        y_trigger = self.target_class
+
+        poison_x, poison_y, poison_index = self.attack.poison(
+            self.data_filepath,
+            x_trigger,
+            y_trigger,
+            x_train,
+            y_train,
+            self.trigger_index,
+        )
+
+        if return_index:
+            return poison_x, poison_y, poison_index
+        return poison_x, poison_y
+
+
+class WitchesBrewScenario(Poison):
+    def load_poisoner(self):
```

Comment:
The `load_poisoner` method seems very long. I think that a lot of this should be passed into the init method of DatasetPoisonerWitchesBrew, if possible.

Reply:
The trick is, the DatasetPoisoner is constructed with an `attack`, and the attack is constructed with these source_class and target_class things.  What do you think about the latest commit, where I just extracted a lot of it into a separate function

Reply:
I like it.